src/NHibernate/Linq/Visitors/ExpressionKeyVisitor.cs

```diff
@@ -86,49 +112,70 @@ protected override Expression VisitConstant(ConstantExpression expression)
 				throw new InvalidOperationException("Cannot visit a constant without a constant to parameter map.");
 			if (_constantToParameterMap.TryGetValue(expression, out param))
 			{
-				// Nulls generate different query plans.  X = variable generates a different query depending on if variable is null or not.
-				if (param.Value == null)
-				{
-					_string.Append("NULL");
-				}
-				else
-				{
-					var value = param.Value as IEnumerable;
-					if (value != null && !(value is string) && !value.Cast<object>().Any())
-					{
-						_string.Append("EmptyList");
-					}
-					else
-					{
-						_string.Append(param.Name);
-					}
-				}
+				VisitParameter(param);
 			}
 			else
 			{
-				if (expression.Value == null)
-				{
-					_string.Append("NULL");
-				}
-				else
-				{
-					var value = expression.Value as IEnumerable;
-					if (value != null  && !(value is string) && !(value is IQueryable))
-					{
-						_string.Append("{");
-						_string.Append(String.Join(",", value.Cast<object>()));
-						_string.Append("}");
-					}
-					else
-					{
-						_string.Append(expression.Value);
-					}
-				}
+				VisitConstantValue(expression.Value);
 			}
 
 			return base.VisitConstant(expression);
 		}
 
+		private void VisitConstantValue(object value)
+		{
+			if (value == null)
+			{
+				_string.Append("NULL");
+				return;
+			}
+
+			if (value is IEnumerable enumerable && !(value is IQueryable))
+			{
+				_string.Append("{");
+				_string.Append(string.Join(",", enumerable.Cast<object>()));
+				_string.Append("}");
+				return;
+			}
+
+			// When MappedAs is used we have to put all sql types information in the key in order to
+			// distinct when different precisions/sizes are used.
+			if (_sessionFactory != null && value is IType type)
```

Comment:
With this PR `ExpressionKeyVisitor` generates different keys when `MappedAs` type is changed in order to avoid using the same query plan as now parameter types are always retrieved from the `ParameterMetadata`.